### PR TITLE
Tenant change related integration tests - allow tenant change and retain tenant in shortlink

### PR DESCRIPTION
**Pull request by leanneeliatra**
### Description

**This integration test covers tests for 2 scenarios:**
1. When a shortlink is copied and later visited, the tenant information is retained in the shortlink and the shortlink is loaded with the correct link/tenant intact. Bug fix for [Bugfix for 1203](https://github.com/opensearch-project/security-dashboards-plugin/issues/1203)
2. After visiting a shared link, the tenant can successfully be changed & the user can continue in the UI. [Bugfix for 1204](https://github.com/opensearch-project/security-dashboards-plugin/issues/1204)

### Issues Resolved
This pull request cover the cypress integration tests for the 2 issues below:
1.  [1203](https://github.com/opensearch-project/security-dashboards-plugin/issues/1203) - When a shortURL link was copied, the tenant information was not being retained & the user was being directed to the incorrect tenant.
2. [1204](https://github.com/opensearch-project/security-dashboards-plugin/issues/1204) - When a page was shared (not using short URL link) & pasted by the user, the user could not then subsequently visit any other tenant. They were locked into the tenant that had previously been copied in the link.

### Questions
How can I see if the integration tests have passed, usually these tests are run in the pipeline when committing to the `security-dashboards-plugin` project. In the instance we push to the test repo. Do I need to consider how the tests will run? Or this repo has it's own pipeline that will take care of checking the tests (which seems to be the case)? Thanks. 

### Check List

- [x] Commits are signed per the DCO using --signoff

### Notes
- The original PR was reviewed by  @cwperks and @DarshitChanpura.
- cc @peternied these are the requested integration tests for [Pull Request](https://github.com/opensearch-project/security-dashboards-plugin/pull/1462) & [Pull Request](https://github.com/opensearch-project/security-dashboards-plugin/pull/1450)
- The original bugfix was merged in pull request [Add the tenant into the short URL once the short URL is resolved #1462
](https://github.com/opensearch-project/security-dashboards-plugin/pull/1462)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+*/
+
+import { CURRENT_TENANT } from '../../../utils/commands';
+import tenantDescription from '../../../fixtures/plugins/security-dashboards-plugin/tenants/testTenant.json';
+
+const tenantName = 'test';
+
+if (Cypress.env('SECURITY_ENABLED')) {
+  describe('Switch tenants when visiting copyed links: ', () => {
```
Comment (cwperks) [inline cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js:15]: nit: copied

File: cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    before(() => {
+      cy.server();
+
+      cy.createTenant(tenantName, tenantDescription);
+    });
+    if (Cypress.browser.name === 'chrome') {
```
Comment (cwperks) [inline cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js:21]: Curious, why is this limited to chrome?

File: cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      it('Checks that the tenant switcher can switch tenants despite a different tenant being present in the tenant query parameter.', function () {
+        CURRENT_TENANT.newTenant = 'private';
+        function switchTenantTo(newTenant) {
```
Comment (cwperks) [inline cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js:24]: Would this function be useful in other tests or is this specific to this test?

File: cypress/integration/plugins/security-dashboards-plugin/tenancy_change_on_shortlink.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+      // 3. Create the short url, wait for response
+      cy.intercept('POST', '/api/shorten_url').as('getShortUrl');
+      // If the url already contains the tenant parameter, it will be stored in the short url. That will work in the app
+      // but would render this test useless. We're testing that resolved short urls without the tenant parameter work as well.
+      cy.url().should('not.contain', 'security_tenant');
+      cy.getElementByTestId('createShortUrl').click();
+      cy.wait('@getShortUrl');
+
+      function switchTenantTo(newTenant) {
```
Comment (cwperks) [inline cypress/integration/plugins/security-dashboards-plugin/tenancy_change_on_shortlink.js:87]: Looks like this function is repeated here? Can this be extracted to a common place?

**Review by cwperks — COMMENTED** [review]
Thank you @leanneeliatra. These tests look good to me. I see a little duplication in the `switchTenantTo` function, could that be extracted into a common location so that its useful in both tests?

File: cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js (hunk 1 above)
Comment (leanneeliatra) [reply cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js]: changed to copied. Thanks

File: cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js (hunk 2 above)
Comment (leanneeliatra) [reply cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js]: That was erroneously left after the testing phase, it has now been removed.

File: cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js (hunk 3 above)
Comment (leanneeliatra) [reply cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js]: It has now been extracted for use in other cypress tests.

File: cypress/integration/plugins/security-dashboards-plugin/tenancy_change_on_shortlink.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      cy.get('[data-test-subj="copyShareUrlButton"]')
+        .invoke('attr', 'data-share-url')
+        .should('contain', '/goto/')
+        .then((shortUrl) => {
+          cy.log('Short url is ' + shortUrl);
+          // Navigate away to avoid the non existing dashboard in the next tenant.
+          // For some reason, using cy.visit() will break things - Cypress can't find the account-popover unless I wait for N seconds.
+          cy.waitForLoader();
+          switchTenantTo('global');
+          // The tests override the cy.visit method, so we need to set the tenant so that the custom command can pick it up.
```
Comment (cwperks) [inline cypress/integration/plugins/security-dashboards-plugin/tenancy_change_on_shortlink.js:64]: I'm confused by this line and the next line. Is the call to `CURRENT_TENANT.newTenant = 'private';` required here? What is the URL that get's navigated to in `cy.visit(shortUrl...`? Is it just the shortUrl or does the call to `CURRENT_TENANT.newTenant` add `?security_tenant=private` to the end of the URL?

File: cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+if (Cypress.env('SECURITY_ENABLED')) {
+  describe('Switch tenants when visiting copied links: ', () => {
+    before(() => {
+      cy.server();
+
+      cy.createTenant(tenantName, tenantDescription);
+    });
+    it('Checks that the tenant switcher can switch tenants despite a different tenant being present in the tenant query parameter.', function () {
+      CURRENT_TENANT.newTenant = tenantName;
```
Comment (ruanyl) [inline cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js:20]: Q: Why changing `defaultTenant` globally? Can we pass `currentTenant` via visit command's `options` to overwrite if `currentTanant` needs to be changed in individual test? Something like this:

```typescript
newOptions.qs = { 
  ...newOptions.qs,
  security_tenant: newOptions.currentTenant ?? CURRENT_TENANT.defaultTenant,
};
```

File: cypress/integration/plugins/security-dashboards-plugin/tenancy_change_on_shortlink.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      cy.intercept({
+        method: 'POST',
+        url: '/api/saved_objects/_bulk_get',
+      }).as('waitForReloadingDashboard');
+      cy.getElementByTestId('confirmSaveSavedObjectButton').click();
+      cy.wait('@waitForReloadingDashboard');
+      cy.wait(2000);
+
+      // 2. Open top share navigation to access copy short url
+      cy.get('[data-test-subj="shareTopNavButton"]').click();
```
Comment (ruanyl) [inline cypress/integration/plugins/security-dashboards-plugin/tenancy_change_on_shortlink.js:43]: Shall we do `cy.getElementByTestId` for consistency?

File: cypress/integration/plugins/security-dashboards-plugin/tenancy_change_on_shortlink.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      // 3. Create the short url, wait for response
+      cy.intercept('POST', '/api/shorten_url').as('getShortUrl');
+      // If the url already contains the tenant parameter, it will be stored in the short url. That will work in the app
+      // but would render this test useless. We're testing that resolved short urls without the tenant parameter work as well.
+      cy.url().should('not.contain', 'security_tenant');
+      cy.getElementByTestId('createShortUrl').click();
+      cy.wait('@getShortUrl');
+
+      //4. Switch tenant & visit shortURL link to ensure tenant from short URL is retained
+      cy.get('[data-test-subj="copyShareUrlButton"]')
```
Comment (ruanyl) [inline cypress/integration/plugins/security-dashboards-plugin/tenancy_change_on_shortlink.js:55]: Shall we do `cy.getElementByTestId` for consistency?

File: cypress/utils/commands.js [hunk 9]
```diff
@@ -46,7 +46,15 @@ Cypress.Commands.overwrite('visit', (orig, url, options) => {
...
     }
-    newOptions.qs = { security_tenant: CURRENT_TENANT.defaultTenant };
+    newOptions.qs = {
+      ...newOptions.qs,
+      security_tenant: CURRENT_TENANT.defaultTenant,
+    };
+
+    if (newOptions.excludeTenant) {
+      delete newOptions.qs.security_tenant;
+    }
```
Comment (ruanyl) [inline cypress/utils/commands.js:56]: It looks we can simply do this?
```suggestion
    if (!newOptions.excludeTenant) {
      newOptions.qs = {
        ...newOptions.qs,
        security_tenant: CURRENT_TENANT.defaultTenant,
      };
    }
```

File: cypress/integration/plugins/security-dashboards-plugin/tenancy_change_on_shortlink.js (hunk 8 above)
Comment (leanneeliatra) [reply cypress/integration/plugins/security-dashboards-plugin/tenancy_change_on_shortlink.js:55]: Sure, thank you!

**Comment by leanneeliatra** [comment]
@cwperks and @ruanyl would it be possible for you to take another look whenever you have a moment please :) Thanks.

File: package.json [hunk 10]
```diff
@@ -7,7 +7,7 @@
     "test": "echo \"Error: no test specified\" && exit 1",
     "cypress:open": "cypress open",
     "cypress:run-without-security": "env TZ=America/Los_Angeles NO_COLOR=1 cypress run --headless --env SECURITY_ENABLED=false",
-    "cypress:run-with-security": "env TZ=America/Los_Angeles NO_COLOR=1 cypress run --headless --env SECURITY_ENABLED=true,openSearchUrl=https://localhost:9200",
```
Comment (cwperks) [inline package.json:10]: Should the `--headless` option remain here?

File: cypress/utils/commands.js [hunk 11]
```diff
@@ -331,7 +337,7 @@ Cypress.Commands.add('deleteSavedObjectByType', (type, search) => {
 Cypress.Commands.add('createIndexPattern', (id, attributes, header = {}) => {
   const url = `${
     Cypress.config().baseUrl
-  }/api/saved_objects/index-pattern/${id}`;
+  }/api/saved_objects/index-pattern/${id}?overwrite=true`; // When running tests locally if ran multiple times the tests fail. The fix is to set Overwrite to true.
```
Comment (cwperks) [inline cypress/utils/commands.js:340]: Is the test not cleaning data after each run? This seems odd to me. A test case should setup and teardown to not affect other test runs.

**Review by DarshitChanpura — APPROVED** [review]

**Review by cwperks — APPROVED** [review]
@leanneeliatra This looks good to merge from my POV. Could a maintainer of this repo please review?

File: cypress/utils/commands.js (hunk 11 above)
Comment (kavilla) [reply cypress/utils/commands.js:340]: Did it fail locally multiple times for the new tests or for all tests? If it was because a previous test failed to succeed and kept stale data since it didn't properly clean up. Then I think we can remove this and ensure before tests cleaned up prior to test run. Otherwise make the overwrite an optional param and default to false incase we wanted to have a test that verifies conflicts can't be added unless it's intended.

File: cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+      cy.visit('/app/home').then(() => {
+        cy.waitForLoader();
+        switchTenantTo('global');
+        cy.waitForLoader();
+        cy.getElementByTestId('account-popover').click();
+        cy.get('#tenantName').should('contain.text', 'Global');
```
Comment (kavilla) [inline cypress/integration/plugins/security-dashboards-plugin/change_tenant_successfully.js:24]: nit: could have a test id as well

File: cypress/integration/plugins/security-dashboards-plugin/switch_tenant.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+export function switchTenantTo(newTenant) {
+  cy.getElementByTestId('account-popover').click();
+  cy.intercept({
+    method: 'GET',
+    url: '/api/v1/auth/dashboardsinfo',
```
Comment (kavilla) [inline cypress/integration/plugins/security-dashboards-plugin/switch_tenant.js:10]: did we every wanted this configurable? or is the url pretty static (`v1`)

Comment (kavilla) [reply cypress/integration/plugins/security-dashboards-plugin/switch_tenant.js:10]: but really thank you for intercepting this, this helps make the test less flaky compared to arbitrary wait times

**Review by kavilla — COMMENTED** [review]
Overall looks pretty good only the overwrite should be address to avoid impacting other tests

File: cypress/utils/commands.js (hunk 11 above)
Comment (cwperks) [reply cypress/utils/commands.js:340]: I'm working on upgrading this repo to Cypress12 now and have noticed a few test suites that are not properly isolated.

Any new PRs should be properly isolating each test case because it will be enforced in the upgraded version of Cypress.

File: cypress/integration/plugins/security-dashboards-plugin/tenancy_change_on_shortlink.js [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+
+      cy.getElementByTestId('savedObjectTitle').type(dashboardName);
+
+      cy.intercept({
+        method: 'POST',
+        url: '/api/saved_objects/_bulk_get',
+      }).as('waitForReloadingDashboard');
+      cy.getElementByTestId('confirmSaveSavedObjectButton').click();
+      cy.wait('@waitForReloadingDashboard');
+      cy.wait(2000);
```
Comment (ashwin-pc) [inline cypress/integration/plugins/security-dashboards-plugin/tenancy_change_on_shortlink.js:94]: nit: If you are already waiting for the request to complete, why the additional 2s?

**Comment by leanneeliatra** [comment]
> Overall looks pretty good only the overwrite should be address to avoid impacting other tests

This is addressed now, thanks for spotting it.

File: cypress/utils/commands.js (hunk 11 above)
Comment (leanneeliatra) [reply cypress/utils/commands.js]: Completed

**Comment by leanneeliatra** [comment]
> Could a maintainer of this repo please review?

Thank you.

**Comment by ruanyl** [comment]
Hi @leanneeliatra Thank you very much for your dedication on the PR! It looks the changed files includes lots of files which are not belong to this PR, could you update your branch accordingly?

**Comment by leanneeliatra** [comment]
> Hi @leanneeliatra Thank you very much for your dedication on the PR! It looks the changed files includes lots of files which are not belong to this PR, could you update your branch accordingly?

Thanks for taking a look @ruanyl, this was due to the fact that the original changes were done on branch 2.8. A rebase to main was then required, that is why there are so many other changes.

I have also resolved conflicts with main just now.

**Comment by leanneeliatra** [comment]
> Hi @leanneeliatra Thank you very much for your dedication on the PR! It looks the changed files includes lots of files which are not belong to this PR, could you update your branch accordingly?

@ruanyl, scene as it would be much clearer, I will add the changes to a new branch, originated from main and just add my changes.

**Comment by leanneeliatra** [comment]
CI is failing on Error as below, can someone assist me on the possible remedy here please? 
```
Path does not exist /home/runner/work/opensearch-dashboards-functional-test/opensearch-dashboards-functional-test/monetery-test/cypress/results
Error: Exit code 1 returned from process: file name '/home/runner/runners/2.309.0/bin/Runner.PluginHost', arguments 'action "GitHub.Runner.Plugins.Artifact.PublishArtifact, Runner.Plugins"'
```
I've been speaking to @Hailong-am. He requested I merge in the main branch, then he could review. Thank you Hailong adding here for clarity on the PR.

**Review by Hailong-am — APPROVED** [review]

File: cypress/integration/common/dashboard_sample_data_spec.js [hunk 15]
```diff
@@ -247,6 +247,7 @@ export function dashboardSanityTests() {
 
       describe('checking discover', () => {
         before(() => {
+          cy.setAdvancedSetting({ 'discover:v2': false });
```
Comment (ruanyl) [inline cypress/integration/common/dashboard_sample_data_spec.js:250]: @leanneeliatra I noticed that you made this change, just want to confirm is this by intention? As this is unrelated to security dashboards plugin.

Comment (leanneeliatra) [reply cypress/integration/common/dashboard_sample_data_spec.js]: Yes, thanks for spotting that ruanyl. I have removed it.

**Comment by leanneeliatra** [comment]
Could a maintainer of this repository please review this PR for me please? Thank you.

**Review by raintygao — APPROVED** [review]

**Comment by ruanyl** [comment]
@kavilla I see E2E Cypress tests are failed, not only this PR, but also others. Looks like OSD server didn't start successfully(can see timeout from the log). Are you familiar with this issue? Did this happen before? I tried to clear the cache , but seems not working.

**Review by ruanyl — APPROVED** [review]

**Comment by ruanyl** [comment]
@leanneeliatra Thanks for the clarification. The PR overall looks good to me, I'll merge it now. And I found this issue https://github.com/opensearch-project/opensearch-dashboards-functional-test/issues/902 which might be related.

**Comment by leanneeliatra** [comment]
> @leanneeliatra Thanks for the clarification. The PR overall looks good to me, I'll merge it now. And I found this issue #902 which might be related. Will track the workflow running issue separately.

That's fabulous @ruanyl thanks very much for your assistance on this. I really appreciate it.

**Comment by ruanyl** [comment]
@ananzh I see this PR is tagged to `v2.10.0`, is that correct?

**Comment by leanneeliatra** [comment]
> @ananzh I see this PR is tagged to `v2.10.0`, is that correct?

It has been now updated to have all the changes from main so it is now updated.
